transformer_engine/pytorch/csrc/extensions.cu

```diff
@@ -1767,6 +1767,174 @@ bool userbuf_comm_available() {  // TODO(ksivamani) check on python side
+            shape[1],
+            shape[2],
+            shape[3],
+            shape[4]);
+    } else {
+        using dtype = at::BFloat16;
+        flash_attention::prepare_kernel<dtype><<<grid, threads, 0,
+                                                 at::cuda::getCurrentCUDAStream()>>>(
+            qkvi.data_ptr<dtype>(),
+            qkv.data_ptr<dtype>(),
+            shape[1],
+            shape[2],
+            shape[3],
+            shape[4]);
+    }
+
+    return qkv;
+}
+
+at::Tensor fa_prepare_bwd(at::Tensor q, at::Tensor k, at::Tensor v) {
+    NVTE_CHECK(q.is_contiguous());
+    NVTE_CHECK(k.is_contiguous());
+    NVTE_CHECK(v.is_contiguous());
+    NVTE_CHECK(q.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(k.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(v.dim() == 4, "Expected 4-dim tensor.");
+    NVTE_CHECK(q.scalar_type() == at::ScalarType::Half ||
+               q.scalar_type() == at::ScalarType::BFloat16);
+    NVTE_CHECK(k.scalar_type() == at::ScalarType::Half ||
+               k.scalar_type() == at::ScalarType::BFloat16);
+    NVTE_CHECK(v.scalar_type() == at::ScalarType::Half ||
+               v.scalar_type() == at::ScalarType::BFloat16);
+    NVTE_CHECK(q.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(q.size(3) == flash_attention::load_size);
+    NVTE_CHECK(k.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(k.size(3) == flash_attention::load_size);
+    NVTE_CHECK(v.size(3) % flash_attention::load_size == 0);
+    NVTE_CHECK(v.size(3) == flash_attention::load_size);
+
+    // 3 x [s, b, n, h] -> [b, s, n, 3 * h]
+
+    std::vector<int64_t> shape = {q.size(1), q.size(0), q.size(2), 3 * q.size(3)};
+    at::Tensor qkv = at::empty(shape, at::CUDA(q.scalar_type()));
+
+    size_t warps = q.size(0) * q.size(1);
+    size_t warps_per_block = flash_attention::block_size / flash_attention::warp_size;
+    size_t blocks = (warps + warps_per_block - 1) / warps_per_block;
+    dim3 grid(blocks, 3);
+    int threads = flash_attention::block_size;
+    if (q.scalar_type() == at::ScalarType::Half) {
+        using dtype = at::Half;
+        flash_attention::prepare_kernel_bwd<dtype><<<grid, threads, 0,
+                                                 at::cuda::getCurrentCUDAStream()>>>(
+            q.data_ptr<dtype>(),
+            k.data_ptr<dtype>(),
+            v.data_ptr<dtype>(),
+            qkv.data_ptr<dtype>(),
+            q.size(0),
+            q.size(1),
+            q.size(2),
+            q.size(3));
+    } else {
+        using dtype = at::BFloat16;
+        flash_attention::prepare_kernel_bwd<dtype><<<grid, threads, 0,
+                                                 at::cuda::getCurrentCUDAStream()>>>(
+            q.data_ptr<dtype>(),
+            k.data_ptr<dtype>(),
+            v.data_ptr<dtype>(),
+            qkv.data_ptr<dtype>(),
+            q.size(0),
+            q.size(1),
+            q.size(2),
+            q.size(3));
+    }
```

Comment:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/Dispatch.h/#L284 seems like a cleaner method

Reply:
Will look into it in a follow up generalization PR (since in principle this code should work for any type, not just FP16 and BF16).